Use Hydra in persistence mode by default: documentation (#7204)
kyma-project/kyma: docs/security/05-05-production-profile.md

---
title: OAuth2 server profiles
type: Configuration
---

> 

Reviewer: An introduction is missing here. Please write a brief introductory sentence describing the profiles we have and when we use them.

## The default profile

By default, every Kyma deployment is installed with the OAuth2 server using what is considered a default profile. In the case of the ORY Hydra OAuth2 server, this means that:

>    - the registered client data is saved in an in-cluster database.

Reviewer: ```suggestion
   - The registered client data is saved in an in-cluster database.
```

   - A job that reads the generated database credentials and saves them to the configuration of Hydra runs before the installation and update.
   - Default resource quotas are used.
   
This configuration is not considered production-ready. To use the Kyma OAuth2 server in a production environment, configure Hydra to use the production profile.

### Persistence mode for the default profile

> The default profile for the OAuth2 server enables the use of a [preconfigured PostgreSQL](https://github.com/helm/charts/tree/master/stable/postgresql) database, which is installed together with the Hydra server. The database is created in the cluster as a StatefulSet and uses a PersistentVolume that is provider-specific. This means that the PersistentVolume used by the database uses the default StorageClass of the cluster's host provider. The internal PostgreSQL database is installed with every Kyma deployment and requires no manual configuration. 

Reviewer: ```suggestion
The default profile for the OAuth2 server enables the use of a [preconfigured PostgreSQL](https://github.com/helm/charts/tree/master/stable/postgresql) database, which is installed together with the Hydra server. The database is created in the cluster as a StatefulSet and uses a PersistentVolume that is provider-specific. This means that the PersistentVolume used by the database uses the default StorageClass of the cluster's host provider. The internal PostgreSQL database is installed with every Kyma deployment and doesn't require manual configuration. 
```

   
## The production profile

The production profile introduces the following changes to the Hydra OAuth2 server deployment:

>    - the registered client data is saved in a user-managed database.

Reviewer: ```suggestion
   - The registered client data is saved in a user-managed database.
```

   - Optionally, a Gcloud proxy service is deployed.

>    - the Oathkeeper authorization and authentication proxy has raised CPU limits and requests. It starts with more replicas and can scale up horizontally to higher numbers.

Reviewer: ```suggestion
   - The Oathkeeper authorization and authentication proxy has raised CPU limits and requests. It starts with more replicas and can scale up horizontally to higher numbers.
```

### Oathkeeper settings for the production profile

The production profile requires the following parameters in order to operate:

| Parameter |  Description | Required value |
|-------|-------|:--------:|
| **oathkeeper.deployment.resources.limits.cpu** | Defines limits for CPU resources. | `800m` |
| **oathkeeper.deployment.resources.requests.cpu** | Defines requests for CPU resources. | `200m` |
| **hpa.oathkeeper.minReplicas** |  Defines the initial number of created Oathkeeper instances. | `3` |

### Persistence modes for the production profile

The production profile for the OAuth2 server enables the use of a custom database, which can be one of the following options:
  - A user-maintained database to which credentials are supplied.
  - A [GCP Cloud SQL](https://cloud.google.com/sql) instance to which credentials are supplied. In this case, an extra [gcloud-proxy](https://cloud.google.com/sql/docs/mysql/sql-proxy) deployment is created allowing secured connections.

#### User-maintained database

> Alternatively, you can use a compatible, custom database to store the registered client data. To use a database, you must create a Kubernetes Secret with the database password as an override for the Hydra OAuth2 server. The details of the database are passed using these parameters of the production profile override:

Reviewer: Shouldn't we change `custom database` into `user-maintained database` here as well?

| **gcloud-sqlproxy.existingSecret** | Specifies the name of the Secret in the same Namespace as the proxy, that stores the database password. | `my-secret` |
| **gcloud-sqlproxy.existingSecretKey** | Specifies the name of the key in the Secret that contains the [GCP ServiceAccount json key](https://cloud.google.com/iam/docs/creating-managing-service-account-keys). | `sa.json` |

>**NOTE:** When using any kind of custom database (gcloud, or self-maintained), it is important to provide the **hydra.hydra.config.secrets** variables, otherwise a random secret will be generated. This secret needs to be common to all Hydra instances using the same instance of the chosen database.

## Use the production profile

> Follow these steps:

Reviewer: Follow these steps to... ? Please elaborate.

1. Apply an override that forces the Hydra OAuth2 server to use the database of your choice. Follow these links to find an example of override data for each persistence mode:

> - [user-maintained](assets/003-ory-db-custom.yaml)

Reviewer: ```suggestion
- [User-maintained](assets/003-ory-db-custom.yaml)
```

> - [Google Cloud SQL](assets/004-ory-db-gcloud.yaml).

Reviewer: ```suggestion
- [Google Cloud SQL](assets/004-ory-db-gcloud.yaml)
```
